File: src/main/java/leetcode/Problem287.java
package leetcode;

public class Problem287 {
    public int findDuplicate(int[] nums) {
        int fast = nums[0];
        int slow = nums[0];
        do {
            fast = nums[nums[fast]];
            slow = nums[slow];
        } while (fast != slow);

        slow = nums[0];
        while (fast != slow) {
            slow = nums[slow];
            fast = nums[fast];
        }
        return slow;
    }
}

File: src/problem_287.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        n = len(nums)
        lo, hi = 1, n
        while lo < hi:
            mi = lo + (hi - lo) // 2
            s = sum(i <= mi for i in nums)
            if s > mi:
                hi = mi
            else:
                lo = mi + 1
        return lo
